core/src/autogluon/core/scheduler/scheduler.py

```diff
@@ -18,6 +21,13 @@
 
 __all__ = ['TaskScheduler']
 
+if mp.get_start_method(allow_none=True) != 'forkserver':
+    # The CUDA runtime does not support the fork start method;
+    # either the spawn or forkserver start method are required to use CUDA in subprocesses.
+    # forkserver is used because spawn is still affected by locking issues
+    logger.warning('WARNING: changing multiprocessing start method to forkserver')
```

Comment:
Is there any drawback to using forkserver?

Reply:
> I know that this is WIP but I'd like to provide the example to help us locate potential issues. I tried the following code with the changes in this PR. But it reports an error:
...
> AttributeError: Can't get attribute 'Net' on <module '__main__' (built-in)>

Every object you want to pass into HPO must be serializable 'on-the-other-side'. Here is how can you re-frame the task:
```python
import numpy as np
import autogluon.core as ag
import torch as th
import torch.nn as nn
import dill

@ag.args(
    num_hidden=ag.space.Int(16, 32),
    lr=ag.space.Real(1e-3, 1e-2)    
)
def train_nn(args, reporter):
    num_hidden = args['num_hidden']
    lr = args['lr']
    batch_size, in_units = 32, 8
    
    class Net(nn.Module):
        def __init__(self, in_units, num_hidden):
            super().__init__()
            self.linear1 = nn.Linear(in_units, num_hidden)
            self.out_layer = nn.Linear(num_hidden, 1)

        def forward(self, x):
            return self.out_layer(nn.ReLU()(self.linear1(x)))
    
    net = Net(in_units=in_units, num_hidden=num_hidden)
    
#     net.cuda()
    net.train()
    optimizer = th.optim.Adam(net.parameters(), lr=lr, amsgrad=True)
    for i in range(10):
        optimizer.zero_grad()
        random_data_in = th.normal(0, 1, (batch_size, in_units))#.cuda()
        random_data_out = 5 * random_data_in ** 2 + 1
        pred_out = net(random_data_in)
        loss = ((random_data_out - pred_out)**2).mean()
        loss.backward()
        with th.no_grad():
#             print(a_gpu)
            print(loss.cpu().detach().numpy())
            reporter(mean_loss=loss.cpu().detach().numpy().item())
        optimizer.step()



scheduler = ag.scheduler.FIFOScheduler(train_nn,
                                   resource={'num_cpus': 4, 'num_gpus': 0},
                                   num_trials=20,
                                   reward_attr='mean_loss',
                                   time_attr='iteration')
scheduler.run()
scheduler.join_jobs()
```
(Have to remove cuda parts since I don't have GPU on my test box)

Reply:
1) more overhead; 2) slower than fork; 3) not supported on all platforms. We enable it only on systems where it is available.
https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods